tests/query_params_test.go

```diff
@@ -0,0 +1,218 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tests
+
+import (
+	"bytes"
+	"context"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-safeweb/testing/requesttesting"
+	"net/http"
+	"net/url"
+	"testing"
+)
+
+const statusOKReq = "HTTP/1.1 200 OK\r\n"
+const statusBadReq = "HTTP/1.1 400 Bad Request\r\n"
+
+// Ensures Query() only returns a map of size one and verifies whether sending a
+// request with two values for the same key returns a []string of length 2
+// containing the correct values
+func TestMultipleQueryParametersSameKey(t *testing.T) {
+	const req = "GET /?vegetable=potatO&vegetable=Tomato HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" +
+		"\r\n"
+	resp, err := requesttesting.MakeRequest(context.Background(), []byte(req), func(req *http.Request) {
+		want := url.Values{
+			"vegetable": []string{"potatO", "Tomato"},
+		}
+		got := req.URL.Query()
+		if diff := cmp.Diff(want, got); diff != "" {
+			t.Errorf("req.URL.Query() = %v, want %v, diff (-want +got):\n%s", got, want, diff)
+		}
+	})
+	if err != nil {
+		t.Fatalf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !bytes.HasPrefix(resp, []byte(statusOKReq)) {
+		t.Errorf("response status: got %s, want %s", resp, statusOKReq)
+	}
+}
+
+func TestQueryParametersSameKeyDifferentCasing(t *testing.T) {
+	req := []byte("GET /?vegetable=potato&Vegetable=tomato HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" +
+		"\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), req, func(req *http.Request) {
+		queryParams := req.URL.Query()
+
+		if len(queryParams) != 2 {
+			t.Errorf("len(queryParams): got %d, want %d", len(queryParams), 2)
+		}
+
+		if len(queryParams["vegetable"]) != 1 || len(queryParams["Vegetable"]) != 1 {
```

Comment:
Why aren't you testing the values? You can use `cmp` then to check the length and values.

Reply:
Resolved in [dca0c50](url)